crates/wmi/src/windows.rs

```diff
@@ -0,0 +1,439 @@
+
+        // SAFETY: Simple FFI call as described in the documentation [1]. This
+        // is based on the official example [2].
+        //
+        // [1]: https://learn.microsoft.com/en-us/windows/win32/api/wbemcli/nf-wbemcli-iwbemservices-execquery
+        // [2]: https://learn.microsoft.com/en-us/windows/win32/wmisdk/example--getting-wmi-data-from-the-local-computer
+        let status = unsafe {
+            use windows_sys::Win32::System::Wmi::*;
+
+            (svc.vtable().ExecQuery)(
+                svc.as_raw_mut(),
+                self::bstr::BString::new("WQL").as_raw_bstr(),
+                self::bstr::BString::new(self.query).as_raw_bstr(),
+                WBEM_FLAG_FORWARD_ONLY | WBEM_FLAG_RETURN_IMMEDIATELY,
+                std::ptr::null_mut(),
+                enum_ptr.as_mut_ptr(),
+            )
+        };
+
+        if status != windows_sys::Win32::Foundation::S_OK {
+            return Err(Error::from_raw_hresult(status).into());
+        }
+
+        Ok(QueryRows {
+            // SAFETY: We verified that the call succeeded, so `enum_ptr` is now
+            // properly initialized and points to a valid WBEM enumerator. Thus,
+            // we can safely construct a wrapper out of it.
+            raw: unsafe {
+                self::com::EnumWbemClassObject::from_raw_ptr(&self.com, enum_ptr.assume_init())
+            },
+        })
+    }
+}
+
+/// Iterator over [rows][1] that a [query][2] yielded.
+///
+/// This struct is created by the [`Query::rows`] method.
+///
+/// [1]: QueryRow
+/// [2]: Query
+pub struct QueryRows<'com> {
+    raw: self::com::EnumWbemClassObject<'com>,
+}
+
+impl<'com> Iterator for QueryRows<'com> {
+
+    type Item = std::io::Result<QueryRow>;
+
+    fn next(&mut self) -> Option<std::io::Result<QueryRow>> {
+        let mut object = match self.raw.next() {
+            None => return None,
+            Some(Ok(object)) => object,
+            Some(Err(error)) => return Some(Err(error)),
+        };
+
+        // SAFETY: We start the enumeration on a valid object without any extra
+        // flags as documented [1]. We verify the call status below.
+        //
+        // [1]: https://learn.microsoft.com/en-us/windows/win32/api/wbemcli/nf-wbemcli-iwbemclassobject-beginenumeration
+        let status = unsafe {
+            (object.vtable().BeginEnumeration)(object.as_raw_mut(), 0)
+        };
+
+        if status != windows_sys::Win32::Foundation::S_OK {
+            return Some(Err(Error::from_raw_hresult(status).into()));
+        }
+
+        let mut row = std::collections::HashMap::new();
+
+        loop {
+            let mut raw_name = std::mem::MaybeUninit::uninit();
+            let mut raw_value = std::mem::MaybeUninit::uninit();
+
+            // SAFETY: We advance the iterator [1] without any extra flags. Note
+            // that we already called `BeginEnumeration`. The call should call
+            // `VariantInit` on `raw_value` [2] so we should not do it upfront.
```

Comment:
The link for [2] seems to be missing

Reply:
Rephrased and removed `[2]` as it refers to content from `[1]` anyway.